server/container/util-java8/src/test/java/org/apache/james/util/FunctionGeneratorTest.java

```diff
@@ -0,0 +1,58 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.util;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.function.Function;
+
+import org.junit.Test;
+
+public class FunctionGeneratorTest {
+
+    @Test
+    public void combineShouldCombine() {
+        FunctionGenerator<Integer, Integer> generateAddition = i -> (j -> j + i);
+
+        Function<Integer, Integer> addSix = generateAddition.combine(generateAddition).apply(3);
+
+        assertThat(addSix.apply(42)).isEqualTo(48);
+    }
+
+    @Test
+    public void combineShouldCombineWithDifferentArguments() {
+        FunctionGenerator<Integer, Integer> generateAddition = i -> (j -> j + i);
+        FunctionGenerator<Integer, Integer> generateDoubleAddition = i -> (j -> j + 2 * i);
+
+        Function<Integer, Integer> addNine = generateAddition.combine(generateDoubleAddition).apply(3);
+
+        assertThat(addNine.apply(42)).isEqualTo(51);
+    }
+
+    @Test
+    public void combineShouldReturnSameResultWhenItCommutes() {
```

Comment:
Yes I know what commuting means :) I just badly read your test title, it is ok :)